### PR TITLE
Reject inconsistent batched particle arrays

### LINKED ISSUE #3449: [BUG] add_particles accepts inconsistent array lengths and produces invalid models
### Bug Description

`ModelBuilder.add_particles()` extends positions, velocities, masses, radii, and flags independently without checking that their lengths match. `finalize()` accepts the resulting builder and creates Warp arrays with different lengths while setting `particle_count` from positions.

A model with two positions and one velocity finalizes successfully. An XPBD CPU step produces non-finite state, while CUDA may read beyond the logical velocity array. Invalid input should fail before mutating the builder, and structural validation should catch manually corrupted per-particle arrays.

### Reproduction Script

```python
import newton

builder = newton.ModelBuilder()
builder.add_particles(
    pos=[(0, 0, 0), (1, 0, 0)],
    vel=[(0, 0, 0)],
    mass=[1.0, 1.0],
)
model = builder.finalize(device="cpu")

print(model.particle_count)
print(model.particle_q.shape)
print(model.particle_qd.shape)
```

Observed output:

```text
2
(2,)
(1,)
```

### System Information

- Newton: current `main` at `9af5a9f4`
- Warp: `1.15.0.dev20260626`

### PR DESCRIPTION
**Pull request by han-xudong**
## Description

Reject inconsistent per-particle input lengths before `ModelBuilder.add_particles()` mutates builder state. Finalization now also validates all particle-frequency arrays against the position-derived particle count unless structural validation is explicitly skipped.

Closes #3449.

## Checklist

- [x] New or existing tests cover these changes
- [x] The documentation is up to date with these changes
- [x] `CHANGELOG.md` has been updated (if user-facing change)

## Test plan

```bash
env WARP_CACHE_ROOT=/tmp/newton-plan04-final-cache uv run --extra dev -m newton.tests -k test_add_particles
env WARP_CACHE_ROOT=/tmp/newton-plan04-validation-cache uv run --extra dev -m newton.tests -k TestModelValidation
uvx pre-commit run -a
git diff --check
```

## Bug fix

**Steps to reproduce:**

1. Call `add_particles()` with two positions and only one velocity or mass.
2. Finalize the builder.
3. Observe a model whose particle arrays have inconsistent lengths, allowing later solvers to read invalid data.

**Minimal reproduction:**

```python
builder.add_particles(
    pos=[(0.0, 0.0, 0.0), (1.0, 0.0, 0.0)],
    vel=[(0.0, 0.0, 0.0)],
    mass=[1.0, 1.0],
)
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened bulk model construction validation to reject inconsistent per-particle input lengths, including cases where required inputs are missing or empty.
  * Prevented inconsistent particle state by enforcing that per-particle arrays remain synchronized during finalization.
* **Tests**
  * Added coverage for mismatched lengths, explicit `None` inputs, empty-position edge cases, and finalization rejection, including a validation-bypass scenario.
* **Documentation / Chores**
  * Updated unreleased changelog notes to reflect the fixes and related entry cleanups.

### CODE REVIEW

File: newton/_src/sim/builder.py [hunk 1]
```diff
@@ -7916,8 +7916,20 @@ def add_particles(
...
+            ("radius", radius),
+            ("flags", flags),
+        )
+        for name, values in particle_inputs:
+            if values is not None and len(values) != particle_count:
+                raise ValueError(
+                    f"{name} length mismatch: expected {particle_count} values to match pos, got {len(values)}"
+                )
+
+        particle_start = self.particle_count
```
[the remark this answers is not in the document]
Comment (han-xudong) [reply newton/_src/sim/builder.py:8194]: Fixed in c4de6d0a. Required `vel` and `mass` inputs now reject `None` before any particle arrays are mutated, while `radius` and `flags` retain their optional broadcasting behavior. Added regression coverage for both missing required inputs and documented the `ValueError` contract in the method docstring. `TestModelValidation`, `test_add_particles`, pre-commit, and `git diff --check` pass.

File: CHANGELOG.md [hunk 2]
```diff
@@ -81,6 +81,7 @@
 ### Fixed
 
 - Fix `ModelBuilder.add_usd()` to honor `ignore_paths` in the custom-frequency traversal, so prims under ignored subtrees no longer register spurious custom-frequency rows in two-pass import workflows. (#3406)
+- Reject inconsistent per-particle array lengths during bulk model construction and finalization. (#3449)
```
Comment (eric-heiden) [inline CHANGELOG.md]: This should point at this PR #3458.

**Review by eric-heiden — COMMENTED** [review]
Thanks @han-xudong, LGTM! Just a small fix is needed for the changelog, then we can merge.

File: CHANGELOG.md (hunk 2 above)
Comment (han-xudong) [reply CHANGELOG.md]: @eric-heiden Thank you for catching this. I updated the changelog reference to #3458 in 830514f8. The targeted test and pre-commit pass.

**Comment by han-xudong** [comment]
@coderabbitai Removed the redundant particle-array length comment in c91944b8; the surrounding validation remains unchanged. Pre-commit passes.

**Review by eric-heiden — APPROVED** [review]
